### PR TITLE
Improve worker status filtering in `hq worker list`

### PR DESCRIPTION
**Pull request by fredykoci**

### CODE REVIEW

**Review by Kobzol — CHANGES_REQUESTED** [review]

File: crates/hyperqueue/src/client/commands/worker.rs [hunk 1]
```diff
@@ -140,24 +166,58 @@ pub async fn start_hq_worker(
     Ok(())
 }
 
+fn lost_worker_info_to_str(worker: &WorkerInfo) -> String {
```
Comment (Kobzol) [inline crates/hyperqueue/src/client/commands/worker.rs:169]: These functions are not needed.

File: crates/hyperqueue/src/client/commands/worker.rs [hunk 2]
```diff
@@ -140,24 +166,58 @@ pub async fn start_hq_worker(
...
+    // }
     if !include_offline {
-        msg.workers.retain(|w| w.ended.is_none());
+        match filter {
+            WorkerFilter::Running => workers.retain(|w| w.ended.is_none()),
+            WorkerFilter::Offline => workers.retain(|w| w.ended.is_some()),
+            WorkerFilter::All => workers.sort_unstable_by_key(|w| w.id),
+        };
+    } else {
+        workers.sort_unstable_by_key(|w| w.id);
```
Comment (Kobzol) [inline crates/hyperqueue/src/client/commands/worker.rs:216]: Let's sort the workers always, before we return them from the function.

File: crates/hyperqueue/src/client/commands/worker.rs [hunk 3]
```diff
@@ -140,24 +166,58 @@ pub async fn start_hq_worker(
...
+    if worker.ended.is_some() {
+        Some(worker.ended.as_ref().unwrap().reason.clone())
+    } else {
+        None
+    }
+}
+
 pub async fn get_worker_list(
     connection: &mut ClientConnection,
     include_offline: bool,
```
Comment (Kobzol) [inline crates/hyperqueue/src/client/commands/worker.rs:188]: `include_offline` is no longer needed, this function should just take `connection` and `filters: Set<WorkerFilter>`.

File: crates/hyperqueue/src/client/commands/worker.rs [hunk 4]
```diff
@@ -20,6 +20,32 @@ use tako::worker::rpc::run_worker;
 use tokio::net::lookup_host;
 use tokio::task::LocalSet;
 
+
+#[derive(Debug)]
+pub enum WorkerFilter {
+    All,
```
Comment (Kobzol) [inline crates/hyperqueue/src/client/commands/worker.rs:26]: Let's remove the `All` variant, it's not needed and it doesn't make sense to use e.g. `--filter all,running`. If the filter set will be empty, all workers will be returned.

**Review by Kobzol — CHANGES_REQUESTED** [review]

File: crates/hyperqueue/src/client/commands/jobs.rs [hunk 5]
```diff
@@ -46,7 +46,14 @@ pub async fn output_job_list(
         response
             .jobs
             .retain(|j| job_filters.contains(&job_status(j)));
+    } else {
+        response.jobs.retain(|j| {
```
[the remark this answers is not in the document]
Comment (fredykoci) [reply crates/hyperqueue/src/client/commands/jobs.rs:50]: this is the reason why this PR starts

File: crates/hyperqueue/src/bin/hq.rs [hunk 6]
```diff
@@ -208,9 +209,18 @@ struct WorkerStopOpts {
...
+    /// Flag to list offline workers
     #[clap(long)]
     all: bool,
+
+    /// Select workers to list [all, running, offline]
+    #[clap(
+        long,
+        require_delimiter(true),
+        use_delimiter(true),
+        multiple_occurrences(false)
```
Comment (Kobzol) [inline crates/hyperqueue/src/bin/hq.rs:221]: These three options are now not needed.

File: crates/hyperqueue/src/bin/hq.rs [hunk 7]
```diff
@@ -208,9 +209,18 @@ struct WorkerStopOpts {
 
 #[derive(Parser)]
 struct WorkerListOpts {
-    /// Include offline workers in the list
+    /// Flag to list offline workers
```
Comment (Kobzol) [inline crates/hyperqueue/src/bin/hq.rs:212]: Please keep the original help message.

File: crates/hyperqueue/src/bin/hq.rs [hunk 8]
```diff
@@ -50,17 +51,17 @@ struct CommonOpts {
...
     /// How should the output of the command be formatted.
-    #[clap(long, env = "HQ_OUTPUT_MODE", default_value = "cli", possible_values = &["cli", "json", "quiet"])]
+    #[clap(long, env = "HQ_OUTPUT_MODE", default_value = "cli", possible_values = & ["cli", "json", "quiet"])]
     output_mode: Outputs,
 }
 
 // Root CLI options
 #[derive(Parser)]
-#[clap(author, about, version(option_env!("HQ_BUILD_VERSION").unwrap_or(env!("CARGO_PKG_VERSION"))))]
+#[clap(author, about, version(option_env ! ("HQ_BUILD_VERSION").unwrap_or(env ! ("CARGO_PKG_VERSION"))))]
```
Comment (Kobzol) [inline crates/hyperqueue/src/bin/hq.rs:64]: Formatting change.

File: crates/hyperqueue/src/client/commands/worker.rs [hunk 9]
```diff
@@ -20,6 +20,28 @@ use tako::worker::rpc::run_worker;
...
+
+impl FromStr for WorkerFilter {
+    type Err = anyhow::Error;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let filter = match input {
+            "running" => WorkerFilter::Running,
+            "offline" => WorkerFilter::Offline,
+            _ => return Err(anyhow::anyhow!(
+                "Invalid worker filter value '{}'. Allowed values are 'all', 'running' or 'offline'.",
```
Comment (Kobzol) [inline crates/hyperqueue/src/client/commands/worker.rs:37]:
```suggestion
                "Invalid worker filter value '{}'. Allowed values are 'running' or 'offline'.",
```

File: crates/hyperqueue/src/client/commands/worker.rs [hunk 10]
```diff
@@ -143,21 +165,32 @@ pub async fn start_hq_worker(
 pub async fn get_worker_list(
     connection: &mut ClientConnection,
     include_offline: bool,
```
Comment (Kobzol) [inline crates/hyperqueue/src/client/commands/worker.rs:167]:
```suggestion

```
This parameter should be removed. If `--all` nor `--filter` is entered, the command code should pass `Some(WorkerFilter::Running)`.

File: crates/hyperqueue/src/transfer/messages.rs [hunk 11]
```diff
@@ -208,7 +208,7 @@ pub struct JobInfo {
 }
 
 // We need to duplicate LostWorkerReason because of serialization problems (msgpack vs. binpack)
-#[derive(Serialize, Deserialize, Debug, Clone)]
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
```
Comment (Kobzol) [inline crates/hyperqueue/src/transfer/messages.rs:211]:
```suggestion
#[derive(Serialize, Deserialize, Debug, Clone)]
```

**Review by Kobzol — CHANGES_REQUESTED** [review]

**Pull request merged**